src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.cs

```diff
@@ -396,11 +403,14 @@ public string GetNameInfo(X509NameType nameType, bool forIssuer)
             if (cchCount == 0)
                 throw Marshal.GetLastWin32Error().ToCryptographicException();
 
-            StringBuilder sb = new StringBuilder(cchCount);
-            if (Interop.crypt32.CertGetNameString(_certContext, certNameType, certNameFlags, ref strType, sb, cchCount) == 0)
-                throw Marshal.GetLastWin32Error().ToCryptographicException();
+            Span<char> buffer = cchCount <= 128 ? stackalloc char[cchCount] : new char[cchCount];
+            fixed (char* ptr = &MemoryMarshal.GetReference(buffer))
+            {
+                if (Interop.crypt32.CertGetNameString(_certContext, certNameType, certNameFlags, ref strType, ptr, cchCount) == 0)
```

Comment:
This chunk of code is repeated 3 times, and the CertGetCertificateContextPropertyString one twice.

Can you make these repeated ones in helper routines in the Interop.crypt32 class so there's less unsafe code / dangerous-if-wrong math floating around?

Reply:
Can you give me an example of what you'd want to see? You want something parameterized with a delegate, passing in an IntPtr and then casting to a pointer (since pointers can be used as generic parmeters), passing all of the arguments in to then be passed into the delegate to avoid a closure, etc.?

Reply:
```C#
internal static unsafe string CertGetNameString(
    SafeCertContextHandle certContext,
    CertNameType certNameType,
    CertNameStrTypeAndFlags strType)
{
    int cchCount = CertGetNameString(certContext, certNameType, certNameFlags, ref strType, null, 0);

    if (cchCount == 0)
    {
        throw Marshal.GetLastWin32Error().ToCryptographicException();
    }

    Span<char> buffer = cchCount <= 128 ? stackalloc char[cchCount] : new char[cchCount];
    fixed (char* ptr = &MemoryMarshal.GetReference(buffer))
    {
        if (CertGetNameString(certContext, certNameType, certNameFlags, ref strType, ptr, cchCount) == 0)
        {
            throw Marshal.GetLastWin32Error().ToCryptographicException();
        }

        Debug.Assert(buffer[cchCount - 1] == 0);
        return new string(buffer.Slice(0, cchCount - 1));
}
```

And then this method is

```C#
public string GetNameInfo(X509NameType nameType, bool forIssuer)
{
    CertNameType certNameType = MapNameType(nameType);
    CertNameFlags certNameFlags = forIssuer ? CertNameFlags.CERT_NAME_ISSUER_FLAG : CertNameFlags.None;
    CertNameStrTypeAndFlags strType = CertNameStrTypeAndFlags.CERT_X500_NAME_STR | CertNameStrTypeAndFlags.CERT_NAME_STR_REVERSE_FLAG;

    return Interop.crypt32.CertGetNameString(_certContext, certNameType, certNameFlags);
}
```

Reply:
Oh, you didn't mean moving all of the boilerplate into a parameterized helper, you meant just moving it into an overload in the interop file.  Sure.

Reply:
Yeah, not factoring out the pattern, but factoring out the literally repeated copies :)

Reply:
Fixed.